relays/substrate/src/headers_initialize.rs

```diff
@@ -88,50 +82,134 @@ async fn do_initialize<SourceChain: Chain, TargetChain: Chain>(
 	Ok(initialization_tx_hash)
 }
 
-/// Prepare initialization data for the headers-bridge pallet.
+/// Prepare initialization data for the finality-verifier pallet.
 async fn prepare_initialization_data<SourceChain: Chain>(
 	source_client: Client<SourceChain>,
-	raw_initial_header: Option<Bytes>,
-	raw_initial_authorities_set: Option<Bytes>,
-	initial_authorities_set_id: Option<GrandpaAuthoritiesSetId>,
 ) -> Result<InitializationData<SourceChain::Header>, String> {
-	let source_genesis_hash = *source_client.genesis_hash();
-
-	let initial_header = match raw_initial_header {
-		Some(raw_initial_header) => SourceChain::Header::decode(&mut &raw_initial_header.0[..])
-			.map_err(|err| format!("Failed to decode {} initial header: {:?}", SourceChain::NAME, err))?,
-		None => source_client
-			.header_by_hash(source_genesis_hash)
-			.await
-			.map_err(|err| format!("Failed to retrive {} genesis header: {:?}", SourceChain::NAME, err))?,
-	};
-
-	let raw_initial_authorities_set = match raw_initial_authorities_set {
-		Some(raw_initial_authorities_set) => raw_initial_authorities_set.0,
-		None => source_client
-			.grandpa_authorities_set(source_genesis_hash)
-			.await
-			.map_err(|err| {
-				format!(
-					"Failed to retrive {} authorities set at genesis header: {:?}",
-					SourceChain::NAME,
-					err
-				)
-			})?,
-	};
-	let initial_authorities_set =
-		GrandpaAuthoritiesSet::decode(&mut &raw_initial_authorities_set[..]).map_err(|err| {
-			format!(
-				"Failed to decode {} initial authorities set: {:?}",
-				SourceChain::NAME,
-				err
-			)
-		})?;
+	// In ideal world we just need to get best finalized header and then to read GRANDPA authorities
+	// set (`pallet_grandpa::CurrentSetId` + `GrandpaApi::grandpa_authorities()`) at this header.
```

Comment:
This should also be solved by https://github.com/paritytech/parity-bridges-common/issues/788

Reply:
Not actually.  The pallet needs to be initialized with set id - otherwise you won't be able to verify most justifications. You can try to find `set_id` by inspecting initial header' ancestors. But reading (possibly many) headers using our public RPC nodes isn't a best idea - I was trying to sync Westend from the genesis on Friday && relay was unable to even submit single transaction - it was able to read ~7000 blocks in 20 minutes, then connection has been lost && it has restarted. So ideally we'll be reading `CurrentSetId`, even if `set_id` will be added to the digest.